### PR TITLE
LightningLite Integration

### LINKED ISSUE #2697: Scale Model Training with LightningLite
**Is your feature/enhancement request related to a problem? Please describe.**
Hi, love this project 💜!
[Trainer](https://github.com/flairNLP/flair/blob/master/flair/trainers/language_model_trainer.py#L160) can be enhanced with support for multiple hardwares and distributed training when training with large dataset or model.

**Describe the solution you'd like**
We have been developing [LightningLite](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html) which  enables you to leverage all the capabilities of PyTorch Lightning Accelerators without any refactoring to your training loop.
You can check out our blog post about it here: [Scale your PyTorch code with LightningLite](https://devblog.pytorchlightning.ai/scale-your-pytorch-code-with-lightninglite-d5692a303f00).

**Additional context**
Overall if you are fine with it I am happy to draft a PR with a suggested change to verify in place the impact. Feel free to get in touch and let me know what you think about this.
Thanks! 😄

### PR DESCRIPTION
**Pull request by aniketmaurya**
Fixes #2697 

This PR attempts to show a refactor to integrate LightningLite for scalable model training with multiple hardware support, mixed precision and DDP.

---
- [x] Ready for Review
- [x] Ready to merge
- [x] Update Example
- [x] Update Tests

### CODE REVIEW

**Comment by aniketmaurya** [comment]
hi @alanakbik, could you pls help in enabling CI tests?

**Comment by alanakbik** [comment]
@aniketmaurya done! just a heads-up: the release of flair 0.11 is imminent, so I'll review the PR after the release!

**Comment by aniketmaurya** [comment]
thanks @alanakbik! Sure, please take your time. I have one request, if you find some time could you help me with a [corpus data](https://github.com/flairNLP/flair/blob/master/resources/docs/TUTORIAL_9_TRAINING_LM_EMBEDDINGS.md#preparing-a-text-corpus) for preparing an example?

**Comment by aniketmaurya** [comment]
> @aniketmaurya done! just a heads-up: the release of flair 0.11 is imminent, so I'll review the PR after the release!

@alanakbik btw we can jointly release this with flair 0.11. I can help with any test required for LightningLite Trainer integration and then we can potentially collaborate on a release article for cross marketing.

This is just an idea. Let me know what you think? 🙂

**Comment by Borda** [comment]
This is awesome! in particular I like that the code is easier to read and reduce some conditional imports :)

File: flair/trainers/language_model_trainer.py
Comment (whoisjones) [inline flair/trainers/language_model_trainer.py:299]: Have you tried running it on single machine - multiple GPUs? I get an error, obviously LightningLite wraps DistributedDataParallel which wraps the LanguageModel class - so this line should be `self.model.module.module.init_hidden`. However this looks not correct to me. When using just CPU, this configurations works. Can you check?

File: flair/trainers/language_model_trainer.py
Comment (whoisjones) [inline flair/trainers/language_model_trainer.py:428]: Same issue as mentioned before.

File: flair/trainers/language_model_trainer.py
Comment (whoisjones) [inline flair/trainers/language_model_trainer.py:434]: If using multiple GPUs, we need to align tensors to be on the same device, I received an issue here.

**Review by whoisjones — COMMENTED** [review]
hi @aniketmaurya, awesome, looking forward to integrate this. Have you tested your development using multiple GPUs? See my comment in the review.

**Comment by Borda** [comment]
> Have you tested your development using multiple GPUs? See my comment in the review.

I think that the next step would be to integrate with EcoCI, which will run nightly also on GPUs
[Stay Ahead of Breaking Changes with the New Lightning Ecosystem CI](https://devblog.pytorchlightning.ai/stay-ahead-of-breaking-changes-with-the-new-lightning-ecosystem-ci-b7e1cf78a6c7)

File: flair/trainers/language_model_trainer.py
Comment (aniketmaurya) [reply flair/trainers/language_model_trainer.py:299]: thanks for checking and reporting this @whoisjones, I let me test this on multiple GPUs and fix it.

File: flair/trainers/language_model_trainer.py
Comment (aniketmaurya) [reply flair/trainers/language_model_trainer.py:436]: `flair.device = self.device` will move the tensor to appropriate device with multiple devices. I have added the fix. :)

**Comment by aniketmaurya** [comment]
Hi @whoisjones, thanks again for reviewing the PR :)

I have fixed the GPU and CPU device mismatch.
For training with DDP strategy we could do conditional check but that won't generalize to other strategies.So we already have started a PR fix for this [here](https://github.com/PyTorchLightning/pytorch-lightning/pull/12597) 😃

**Review by Borda — APPROVED** [review]
lgtm

**Comment by aniketmaurya** [comment]
> Hey @aniketmaurya have you already heard some from pytorch-lightning community?

Hi @whoisjones, sorry didn't get you :)

**Comment by alanakbik** [comment]
Hi @aniketmaurya I think the question is about this PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/12597 - if I understand correctly, it needs to be merged first so that LightningLite runs for Flair with DDP strategy.

**Comment by aniketmaurya** [comment]
Hi @alanakbik, sorry for the late response!
Yes, the model (`nn.Module`) is wrapped into the `DistributedDataParallel` class so we are not able to access the `init_hidden` method directly but we can use a conditional statement to get there, until [this PR](https://github.com/PyTorchLightning/pytorch-lightning/pull/12597) is merged. Although this would be a hack around and won't generalize with other strategies available in PyTorch Lightning like `Deepspeed` and `ddp_spwan`.

**Comment by alanakbik** [comment]
Thanks for the quick fix @aniketmaurya! Its probably best though to wait until the PR is merged as otherwise it might be confusing for users that only a subset of PyTorch Lightning capabilities is available.

**Comment by machinelearnear** [comment]
Thanks everyone for the great work on this, I've been trying to get Flair to work on single instance multi-gpu as well, so very interested in this PR. I see that the [Lighting PR](https://github.com/PyTorchLightning/pytorch-lightning/pull/12597) was finally merged 7 days ago.

**Comment by aniketmaurya** [comment]
Hi @machinelearnear, thank you for your support!
@alanakbik, Lightning has also deprecated Python 3.6. We will have to get the latest updated version with [bug fixes](https://github.com/PyTorchLightning/pytorch-lightning/pull/12597) by upgrading from Python 3.6.

I already have [this PR](https://github.com/flairNLP/flair/pull/2775), which removes Python 3.6 for Flair.

**Comment by alanakbik** [comment]
@aniketmaurya that is great, thanks for the update! We will do another minor release first and reserve the update to Python 3.7 for the next bigger release - I'll keep you posted!

**Comment by machinelearnear** [comment]
@aniketmaurya any plans to also integrate `trainer.py` next? I did some tests here: https://github.com/machinelearnear/use-lightninglite-sagemaker-and-flair/blob/main/example-flair/notebook_flair.ipynb but still having problems porting some functions to PL (batches are split differently across GPUs for some reason). Thanks for all the nice work!

**Comment by dchaplinsky** [comment]
Any updates on this one? I have two GPUs now and eager to try it with Flair!

**Comment by Borda** [comment]
cc: @awaelchli :rabbit:

**Comment by alanakbik** [comment]
@dchaplinsky I tried today with the most recent lightning release and DDP strategy, using the code from the example that comes with the PR: 

```python
trainer = LanguageModelTrainer(accelerator="gpu", devices="auto", strategy='ddp')
```

It worked quite well in that it was automatically using 4 GPUs that were all going at 100%. It seems each GPU gets a different split of the data. 

However, when epoch 1 ended, it got stuck without throwing an error, and I terminated execution after a while. @aniketmaurya any idea why this might be the case?

**Comment by aniketmaurya** [comment]
hi @alanakbik, Is it possible to get the data and training details to reproduce this error? Also could you please provide the error trace?

**Comment by alanakbik** [comment]
Hello @aniketmaurya sure, I'll build a minimal example for you.

**Comment by alanakbik** [comment]
Here is a minimal training data example for my training script. Unpack this where you like and point the script to the root folder:
[penn_lm.zip](https://github.com/flairNLP/flair/files/9899426/penn_lm.zip). Note that we usually split the training data into a folder with many smaller splits. The data loader can then load the splits asynchonously. Patience is computed against the splits which allows for annealing mid-epoch.  

I used the following training script on a machine with 3 Nividia 3090 GPUs and Python 3.8: 

```python
import flair

from flair.data import Dictionary
from flair.models import LanguageModel
from flair.trainers.language_model_trainer import LanguageModelTrainer, TextCorpus

flair.device = 'cuda:0'

# are you training a forward or backward LM?
is_forward_lm = True

# get your corpus, process forward and at the character level
dictionary = Dictionary.load('chars')

# get corpus
corpus = TextCorpus("resources/tasks/penn_lm/",
                    dictionary,
                    is_forward_lm,
                    character_level=True,
                    random_case_flip=True,
                    )

language_model = LanguageModel(dictionary,
                               is_forward_lm=True,
                               hidden_size=128,
                               nlayers=1,
                               )
print(language_model)

# train your language model
trainer = LanguageModelTrainer(accelerator="gpu", devices="auto", strategy='ddp')

# train your language model
trainer.train(language_model,
              corpus,
              "resources/LMs/penn-gpu-auto-ddp",
              sequence_length=50,
              mini_batch_size=10,
              learning_rate=10,
              patience=10,
              max_epochs=20,
              checkpoint=True,
              num_workers=4,
              )

print(language_model.generate_text())
```

It automatically takes all three GPUs and runs them at 100% (great), but at the end of the first epoch nothing more happens. Here is the last output I get: 

```console
2022-10-31 09:34:56,368 read text file with 2000 lines
2022-10-31 09:34:56,369 shuffled
2022-10-31 09:34:56,372 Sequence length is 50
2022-10-31 09:34:56,373 Split 1  - (09:34:56)
2022-10-31 09:34:56,379 read text file with 2000 lines
2022-10-31 09:34:56,380 shuffled
2022-10-31 09:34:56,382 read text file with 2000 lines
2022-10-31 09:34:56,383 shuffled
2022-10-31 09:34:56,384 read text file with 2000 lines
2022-10-31 09:34:56,385 shuffled
2022-10-31 09:34:56,386 read text file with 2000 lines
2022-10-31 09:34:56,386 Sequence length is 50
2022-10-31 09:34:56,386 shuffled
2022-10-31 09:34:56,387 Split 1  - (09:34:56)
2022-10-31 09:34:56,395 read text file with 2000 lines
2022-10-31 09:34:56,396 read text file with 2000 lines
2022-10-31 09:34:56,396 shuffled
2022-10-31 09:34:56,396 shuffled
2022-10-31 09:34:56,399 read text file with 2000 lines
2022-10-31 09:34:56,399 shuffled
2022-10-31 09:34:56,419 read text file with 42068 lines
2022-10-31 09:34:56,436 shuffled
2022-10-31 09:34:59,746 | split   1/ 23 |   100/  483 batches | ms/batch 33.72 | loss 3.0842 | ppl 21.8503
2022-10-31 09:34:59,746 | split   1/ 23 |   100/  478 batches | ms/batch 33.59 | loss 3.0945 | ppl 22.0756
2022-10-31 09:34:59,746 | split   1/ 23 |   100/  480 batches | ms/batch 34.54 | loss 3.1399 | ppl 23.1021
2022-10-31 09:34:59,927 | split   1/ 23 |   200/  483 batches | ms/batch  1.80 | loss 2.2430 | ppl 9.4220
2022-10-31 09:34:59,927 | split   1/ 23 |   200/  478 batches | ms/batch  1.80 | loss 2.2366 | ppl 9.3619
2022-10-31 09:34:59,927 | split   1/ 23 |   200/  480 batches | ms/batch  1.80 | loss 2.2417 | ppl 9.4089
2022-10-31 09:35:00,102 | split   1/ 23 |   300/  483 batches | ms/batch  1.75 | loss 2.0134 | ppl 7.4888
2022-10-31 09:35:00,102 | split   1/ 23 |   300/  478 batches | ms/batch  1.75 | loss 2.0211 | ppl 7.5464
2022-10-31 09:35:00,102 | split   1/ 23 |   300/  480 batches | ms/batch  1.75 | loss 2.0382 | ppl 7.6765
2022-10-31 09:35:00,278 | split   1/ 23 |   400/  480 batches | ms/batch  1.76 | loss 1.8816 | ppl 6.5641
2022-10-31 09:35:00,278 | split   1/ 23 |   400/  478 batches | ms/batch  1.76 | loss 1.8837 | ppl 6.5780
2022-10-31 09:35:00,278 | split   1/ 23 |   400/  483 batches | ms/batch  1.77 | loss 1.8589 | ppl 6.4168
```

On a single GPU on my local machine it works fine.

**Comment by alanakbik** [comment]
Quick update: it trains successfully if the `train` folder only contains a single split, so maybe the data loader messes up the execution.

**Comment by dchaplinsky** [comment]
I've tried this branch on my 2 GPU (2060 and 1070) setup while training
flair embeddings for Ukrainian (using different strategies, half of which
didn't work :(). Unfortunately in this setup it gave me no performance
boost so I switched to a different scheme: trained forward and backward
embeddings on different GPUs (for a month+)

Would be happy to try it again one day.

**Comment by alanakbik** [comment]
@dchaplinsky form what I can tell, it does a mini-batch on each GPU and then gathers the gradients for a single update. So in the same amount of time, you double (with two GPUs) your mini-batch size. So this seems to be primarily to achieve higher mini-batch sizes by using multiple GPUs. 

I think there is currently something wrong in the way the splits are distributed across GPUs. I would expect the dataset to be partitioned so that each GPU to get different splits. But currently it looks like each GPU gets the full data set, which is why there is no speedup.

**Comment by dchaplinsky** [comment]
Yeah, similar feelings. I ended up modifying the code for LM trainer,
implementing a poor man version of grad accum and ramped up batch 5 times.

**Comment by dchaplinsky** [comment]
Please keep it open and make it happen!

**Comment by Borda** [comment]
@aniketmaurya could you update it with 2.0? 🐿️

**Comment by aniketmaurya** [comment]
created a new [PR](https://github.com/flairNLP/flair/pull/3159) with Fabric (revamped LightningLite). It is much cleaner!!

**Pull request closed without merging**
